egs/wsj/s5/local/nnet3/run_lstm.sh

```diff
@@ -0,0 +1,127 @@
+#!/bin/bash
+
+# this is a basic lstm script
+# LSTM script runs for more epochs than the TDNN script
+# and each epoch takes twice the time
+
+# At this script level we don't support not running on GPU, as it would be painfully slow.
+# If you want to run without GPU you'd have to call lstm/train.sh with --gpu false
+
+stage=0
+train_stage=-10
+affix=
+common_egs_dir=
+
+# LSTM options
+splice_indexes="-2,-1,0,1,2 0 0"
+label_delay=5
+num_lstm_layers=3
+cell_dim=1024
+hidden_dim=1024
+recurrent_projection_dim=256
+non_recurrent_projection_dim=256
+chunk_width=20
+chunk_left_context=40
+clipping_threshold=30.0
+norm_based_clipping=true
+
+num_epochs=10
+
+# training options
+initial_effective_lrate=0.0003
+final_effective_lrate=0.00003
+num_jobs_initial=2
+num_jobs_final=12
+momentum=0.5
+shrink=0.98
```

Comment:
I see. You are right, it'll compile to a cyclic graph.
Then I should go for the --lstm-delay "-1 1 -1 1" plan.
In that way, lstm2 should take input from "input" other than "lstm1", so 
that lstm1 and lstm2 make a bidirectional layer.
Then, lstm3 takes input from both lstm1 and lstm2, such as
```
component-node name=Lstm3_i1 component=Lstm3_W_i-xr 
input=Append(Lstm1_rp_t, Lstm2_rp_t, IfDefined(Offset(Lstm3_r_t, -1)))
```
Does it make sense?

On 10/16/2015 01:24 PM, Daniel Povey wrote:
>
> In egs/wsj/s5/local/nnet3/run_lstm.sh 
> <https://github.com/kaldi-asr/kaldi/pull/256#discussion_r42209628>:
>
> > +recurrent_projection_dim=256
> > +non_recurrent_projection_dim=256
> > +chunk_width=20
> > +chunk_left_context=40
> > +clipping_threshold=30.0
> > +norm_based_clipping=true
> > +
> > +num_epochs=10
> > +
> > +# training options
> > +initial_effective_lrate=0.0003
> > +final_effective_lrate=0.00003
> > +num_jobs_initial=2
> > +num_jobs_final=12
> > +momentum=0.5
> > +shrink=0.98
> Xingyu, I think a bidirectional layer like that won't be definable, as 
> it will compile to a cyclic graph, and the tool will just reject it. 
> Unless I'm missing something, there is no way to interpret something 
> like that as a forward type of computation.
> … <#>
> On Fri, Oct 16, 2015 at 1:19 AM, Xingyu Na <notifications@github.com> 
> wrote: In egs/wsj/s5/local/nnet3/run_lstm.sh <#256 (comment) 
> <https://github.com/kaldi-asr/kaldi/pull/256#discussion_r42209446>>: > 
> +recurrent_projection_dim=256 > +non_recurrent_projection_dim=256 > 
> +chunk_width=20 > +chunk_left_context=40 > +clipping_threshold=30.0 > 
> +norm_based_clipping=true > + > +num_epochs=10 > + > +# training 
> options > +initial_effective_lrate=0.0003 > 
> +final_effective_lrate=0.00003 > +num_jobs_initial=2 > 
> +num_jobs_final=12 > +momentum=0.5 > +shrink=0.98 I'm trying defining 
> a bidirectional layer with (-1, +1) other than defining two single 
> directional layers with (-1) and (1). It requires the gate 
> component-node to sum three components, such as: ``` component-node 
> name=Lstm1_if1 component=Lstm1_W_i-xr input=Append(L0_lda, 
> IfDefined(Offset(Lstm1_r_t, -1))) component-node name=Lstm1_if2 
> component=Lstm1_w_ic input=IfDefined(Offset(Lstm1_c_t, -1)) 
> component-node name=Lstm1_ib1 component=Lstm1_W_i-xr 
> input=Append(L0_lda, IfDefined(Offset(Lstm1_r_t, 1))) component-node 
> name=Lstm1_ib2 component=Lstm1_w_ic input=IfDefined(Offset(Lstm1_c_t, 
> 1)) component-node name=Lstm1_i_t component=Lstm1_i 
> input=Sum(Lstm1_if1, Lstm1_if2, Lstm1_ib1, Lstm1_ib2) ``` This is not 
> like the one in Alex (ASRU2013). Or, if I use the current lstm maker, 
> with ```--lstm-delay "-1 1 -1 1"``` and some extra right context, I 
> still need to pass the output of the first layer to the input of the 
> third layer, right? Xingyu … <#1506f167d5d42769_> On 10/16/2015 01:54 
> AM, Daniel Povey wrote: In egs/wsj/s5/local/nnet3/run_lstm.sh <#256 
> <https://github.com/kaldi-asr/kaldi/pull/256> (comment) <#256 
> (comment) 
> <https://github.com/kaldi-asr/kaldi/pull/256#discussion_r42157224>>>: 
> > +recurrent_projection_dim=256 > +non_recurrent_projection_dim=256 > 
> +chunk_width=20 > +chunk_left_context=40 > +clipping_threshold=30.0 > 
> +norm_based_clipping=true > + > +num_epochs=10 > + > +# training 
> options > +initial_effective_lrate=0.0003 > 
> +final_effective_lrate=0.00003 > +num_jobs_initial=2 > 
> +num_jobs_final=12 > +momentum=0.5 > +shrink=0.98 Guys, some comments 
> on this-- - If convergence is slow, you could just increase the 
> learning rates, assuming you are using all the changes in the current 
> master (e.g. relating to how shrinkage is applied, and the 
> max-change). We don't have divergence problems now. - I don't 
> understand why you need a separate set of scripts for BLSTMs. Isn't it 
> just a question of putting some positive values in the --lstm-delay 
> option, and adding an --extra-right-context option to the script? … 
> <#> On Thu, Oct 15, 2015 at 1:22 PM, Yiming Wang 
> <notifications@github.com> wrote: In 
> egs/wsj/s5/local/nnet3/run_lstm.sh <#256 
> <https://github.com/kaldi-asr/kaldi/pull/256> <#256 
> <https://github.com/kaldi-asr/kaldi/pull/256>> (comment) <#256 
> <https://github.com/kaldi-asr/kaldi/pull/256> (comment) <#256 
> (comment) 
> <https://github.com/kaldi-asr/kaldi/pull/256#discussion_r42153253>>>>: 
> > +recurrent_projection_dim=256 > +non_recurrent_projection_dim=256 > 
> +chunk_width=20 > +chunk_left_context=40 > +clipping_threshold=30.0 > 
> +norm_based_clipping=true > + > +num_epochs=10 > + > +# training 
> options > +initial_effective_lrate=0.0003 > 
> +final_effective_lrate=0.00003 > +num_jobs_initial=2 > 
> +num_jobs_final=12 > +momentum=0.5 > +shrink=0.98 I checked the latest 
> swbd lstm recipe in the master branch, but the defalut shrink value is 
> still 0.98? — Reply to this email directly or view it on GitHub 
> <https://github.com/kaldi-asr/kaldi/pull/256/files#r42153253>. — Reply 
> to this email directly or view it on GitHub < 
> https://github.com/kaldi-asr/kaldi/pull/256/files#r42157224>. — Reply 
> to this email directly or view it on GitHub 
> <https://github.com/kaldi-asr/kaldi/pull/256/files#r42209446>.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/kaldi-asr/kaldi/pull/256/files#r42209628>.
>